pkg/metrics/metrics.go

```diff
@@ -0,0 +1,164 @@
+package metrics
+
+import (
+	"context"
+	"github.com/kubeslice/kubeslice-monitoring/pkg/util"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/runtime"
+	"time"
+)
+
+// MetricRecorder provides a container with config parameters for the Prometheus Exporter
+type MetricRecorder struct {
+	Logger    *zap.SugaredLogger
+	Scheme    *runtime.Scheme
+	Version   string
+	Project   string
+	Cluster   string
+	Slice     string
+	Namespace string
+	Subsystem string
+}
+
+type Metric struct {
+	Type             MetricType
+	Name             string
+	Help             string
+	Value            float64
+	Labels           map[string]string
+	histogramBuckets []float64
+}
+
+type MetricType string
+
+var fixedLabels = []string{"slice_project", "slice_cluster", "slice_name"}
+
+const (
+	MetricTypeGauge     = "Gauge"
+	MetricTypeCounter   = "Counter"
+	MetricTypeHistogram = "Histogram"
+	MetricTypeSummary   = "Summary"
+)
+
+// NewPrometheusProvider returns a Provider that produces Prometheus metrics.
+// Namespace and subsystem are applied to all produced metrics.
+func (mr *MetricRecorder) NewPrometheusProvider() *MetricRecorder {
+	return &MetricRecorder{
+		Logger:    mr.Logger,
+		Scheme:    mr.Scheme,
+		Version:   mr.Version,
+		Project:   mr.Project,
+		Cluster:   mr.Cluster,
+		Slice:     mr.Slice,
+		Namespace: mr.Namespace,
+		Subsystem: mr.Subsystem,
+	}
+}
+
+func (mr *MetricRecorder) RecordMetric(ctx context.Context, m *Metric) error {
+	switch m.Type {
+	case MetricTypeGauge:
+		mr.recordGauge(ctx, m)
+		break
+	case MetricTypeCounter:
+		mr.recordCounter(ctx, m)
+		break
+	case MetricTypeHistogram:
+		mr.recordHistogram(ctx, m)
+		break
+	case MetricTypeSummary:
+		mr.recordSummary(ctx, m)
+		break
+	default:
+		mr.Logger.Errorf("Unknown metric type: %v", m.Type)
+	}
+	return nil
+}
+
+func (mr *MetricRecorder) recordGauge(ctx context.Context, m *Metric) {
+	mr.Logger.Infof("Recording gauge metric: %v", m)
```

Comment:
I don't think we should log this in info mode. Metric recording can happen very frequently and it is of no use to normal users

Reply:
Changed it to debug.